UrlForm and AWS query support

dhpiggott opened the pull request.

dhpiggott committed:
UrlForm support

dhpiggott committed:
Wire up codec, client and compliance tests

dhpiggott committed:
Fix tests... except the one that seems like it requires a remodel

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/disneystreaming/smithy4s?pullRequest=1101) <br/>All committers have signed the CLA.

dhpiggott committed:
Disgusting hacks

Obviously we don't want this to be the permanent solution but it at
least confirms understanding of the problem.

With this in place all the compliance tests except for the idempotency
token case (which we're ignoring for now anyway) pass.

dhpiggott committed:
Less stinky solve for collection/map encoding difference

Credit to Oli.

dhpiggott committed:
Smoke test the AWS query protocol by listing CloudWatch metrics

kubukoz commented on modules/aws-http4s/src/smithy4s/aws/AwsCredentialsProvider.scala:
```diff
@@ -62,6 +62,8 @@ class AwsCredentialsProvider[F[_]](implicit F: Temporal[F]) {
     Resource
       .eval(fromEnv)
       .map(F.pure)
+      // TODO: Ensure minimal delay when these endpoints don't exist, e.g.
+      // when running locally.
```
related: https://github.com/disneystreaming/smithy4s/issues/1075

dhpiggott commented:
Closing this because it includes unfinished support for encoding and decoding - whereas the plan is to get encoding support good enough to merge and then follow up with decoding support. So I'll open a new PR for encoding support and then a follow up will add decoding support.

dhpiggott commented:
Encoding support draft: https://github.com/disneystreaming/smithy4s/pull/1112

dhpiggott commented:
Decoding support will be added back after that in https://github.com/disneystreaming/smithy4s/pull/1113.
